cmd/src/validate.go

```diff
@@ -0,0 +1,609 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/ghodss/yaml"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/mattn/go-isatty"
+)
+
+type validationSpec struct {
+	FirstAdmin struct {
+		Email    string
+		Username string
+		Password string
+	}
+	WaitRepoCloned struct {
+		Repo                     string
+		MaxTries                 int
+		SleepBetweenTriesSeconds int
+	}
+	SearchQuery     string
+	ExternalService struct {
+		Kind           string
+		DisplayName    string
+		Config         *json.RawMessage
+		DeleteWhenDone bool
+	}
+}
+
+type validator struct {
+	client *vdClient
+}
+
+func init() {
+	usage := `'src validate' is a tool that validates a Sourcegraph instance.
+
+EXPERIMENTAL: 'validate' is an experimental command in the 'src' tool.
+
+Usage:
+
+	src validate [options] <script-file>
+or
+    cat <script-file> | src validate [options]
+`
+	flagSet := flag.NewFlagSet("validate", flag.ExitOnError)
+	usageFunc := func() {
+		_, _ = fmt.Fprintf(flag.CommandLine.Output(), "Usage of 'src validate %s':\n", flagSet.Name())
+		flagSet.PrintDefaults()
+		fmt.Println(usage)
+	}
+
+	var (
+		contextFlag = flagSet.String("context", "", `Comma-separated list of key=value pairs to add to the script execution context`)
+		docFlag     = flagSet.Bool("doc", false, `Show documentation`)
+		secretsFlag = flagSet.String("secrets", "", "Path to a file containing key=value lines. The key value pairs will be added to the script context")
+	)
+
+	vd := &validator{}
+
+	commands = append(commands, &command{
+		flagSet: flagSet,
+		handler: func(args []string) error {
+			if *docFlag {
+				vd.printDocumentation()
+				return nil
+			}
+
+			var script []byte
+			var err error
+			if len(flagSet.Args()) == 1 {
+				filename := flagSet.Arg(0)
+				script, err = ioutil.ReadFile(filename)
+				if err != nil {
+					return err
+				}
+				if strings.HasSuffix(filename, ".yaml") || strings.HasSuffix(filename, ".yml") {
+					script, err = yaml.YAMLToJSONStrict(script)
+					if err != nil {
+						return err
+					}
+				}
+			}
+			if !isatty.IsTerminal(os.Stdin.Fd()) {
+				// stdin is a pipe not a terminal
+				script, err = ioutil.ReadAll(os.Stdin)
+				if err != nil {
+					return err
+				}
+				script, err = yaml.YAMLToJSONStrict(script)
+				if err != nil {
+					return err
+				}
+			}
+
+			ctxm := vd.parseKVPairs(*contextFlag, ",")
+
+			if *secretsFlag != "" {
+				sm, err := vd.readSecrets(*secretsFlag)
+				if err != nil {
+					return err
+				}
+
+				for k, v := range sm {
+					ctxm[k] = v
+				}
+			}
+
+			return vd.validate(script, ctxm)
+		},
+		usageFunc: usageFunc,
+	})
+}
+
+func (vd *validator) printDocumentation() {
+	fmt.Println("Please visit https://docs.sourcegraph.com/admin/validation for documentation of the validate command.")
```

Comment:
Please add docs for this command to the handbook under distribution (or the actual docs if we want it public). Specifically the config format being documented would be nice.